### ISSUE TITLE
Exception when copying fits HDU after reading then writing it

### ISSUE THREAD

**Issue by brgillis** [astropy/astropy#5407]
I've found what seems to be a bug which occurs after you open a fits image, write it, and then copy the image HDU. I can trigger it using the following script:

```
from astropy.io.fits import HDUList, PrimaryHDU, open
from copy import deepcopy
import numpy as np

hdulist = HDUList([PrimaryHDU(np.zeros((10,10)))])

hdulist.writeto("junk",clobber=True)

newhdulist = open("junk")

newhdulist.writeto("junk",clobber=True)

im = deepcopy(newhdulist[0])
```

This gives the Traceback:

```
Traceback (most recent call last):
  File "crash_test.py", line 13, in <module>
    im = deepcopy(newhdulist[0])
  File "/usr/lib64/python2.7/copy.py", line 190, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/usr/lib64/python2.7/copy.py", line 334, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib64/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/usr/lib64/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib64/python2.7/copy.py", line 190, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/usr/lib64/python2.7/copy.py", line 334, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib64/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/usr/lib64/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib64/python2.7/copy.py", line 190, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/usr/lib64/python2.7/copy.py", line 329, in _reconstruct
    y = callable(*args)
  File "/usr/lib64/python2.7/copy_reg.py", line 93, in __newobj__
    return cls.__new__(cls, *args)
```

There seem to be some odd implementation-dependent things going on here as well. With astropy v1.2.1, it crashes on a cluster I'm working on, but not my desktop (using Python v2.7.9 for both). With astropy v1.0.10 though, it crashed on both.

**Comment by pllim** [comment]
Are you actually trying to just make a deep copy of the *data* or the entire `PrimaryHDU`? Naming the variable `im` seems to suggest that you actually want `im = deepcopy(newhdulist[0].data)` (which works fine).

Assuming that you want to actually copy the HDU, on Python 3.5 and latest Astropy dev, I see `TypeError: cannot serialize '_io.FileIO' object`. However, `im = newhdulist[0].copy()` works fine in that case.

**Comment by brgillis** [comment]
I do want the full data; that was just poor naming on my part. What first caused problems for me was when I was simply returning (the equivalent of) newhdulist[0] from a function which I was running in parallel (using multiprocessing.Pool). The clean-up at the end of its execution triggered this error, I assume because it tried to copy the HDU I'd returned.

I found a workaround that works for my purposes: `im = fits.PrimaryHDU(newhdulist[0].data,newhdulist[0].header)`, but this is non-obvious.

**Comment by pllim** [comment]
I am not sure if there is an actual bug here. Perhaps just a special use case? If you think this is no longer an issue, feel free to close.

**Comment by brgillis** [comment]
I think this is a bug, albeit one that only shows up in a corner case (and so it's probably not going to be a high priority to fix). There's no obvious reason from the user's standpoint why there might be a problem copying an HDU in these circumstances (it only shows up after loading from a file and then writing to one; either step alone doesn't seem to trigger it).

**Comment by kakirastern** [comment]
Hi @brgillis, I tried the following instead and it seems okay... 
```
from astropy.io.fits import HDUList, PrimaryHDU, open
from copy import deepcopy
import numpy as np

hdulist = HDUList([PrimaryHDU(np.zeros((10,10)))])

hdulist.writeto("junk",clobber=True)

newhdulist = open("junk")

newhdulist.writeto("junk",clobber=True)

im = deepcopy(newhdulist)[0]
```
This should produce an `im` as `<astropy.io.fits.hdu.image.PrimaryHDU at 0x1149dc390>`. 

But if I use your code with the last line un-modified with Astropy v3.1.1, then I would get an error thrown as `TypeError: can't pickle mmap.mmap objects`. I think there has been some idiosyncrasy with the syntax you have used.

**Comment by pllim** [comment]
This issue is pretty old and was using Python 2 that is no longer supported. Nor was it using context manager. I am going to close this. If you still run into problem with Python 3, using context manager, and using Astropy dev, then please open a new issue. Thanks!
